File: src/js/core/diagram/code/evalexprstencil.js
define([
  "core/expr/code/impl",
  "core/expr/code/env",
  "core/diagram/code/traceval",
  "core/schema/code/factory",
  "core/system/library/schema"
],
function(Impl, Env, Traceval, Factory, Schema) {
  var Evalexprstencil ;
  var EvalExprStencil = MakeMixin([Traceval.TracevalCommand], function() {
    this.eval_Rule = function(obj) {
      var self = this; 
      var funname, forms;
      funname = S(obj.name(), "__", obj.type());
      forms = [obj.obj()];
      obj.formals().each(function(f) {
        return forms.push(f.name());
      });
      self.$.D._get("env")._set(funname, Impl.Closure.make_closure(obj.body(), forms, self.$.D._get("env"), self));
      return self.$.D._get("srctemp")._set(funname, Impl.Closure.new(obj.body(), forms, self.$.D._get("env"), self));
    };

    this.eval_RuleCall = function(obj) {
      var self = this; 
      var target, funname, m, params, clos, newsrctmp, param_src, res;
      target = self.eval(obj.obj());
      funname = S(obj.name(), "__", target.schema_class().name());
      m = self.$.D._get("env")._get(funname);
      params = obj.params().map(function(p) {
        return self.eval(p);
      });
      clos = self.$.D._get("srctemp")._get(funname);
      newsrctmp = self.$.D._get("srctemp").clone();
      clos.formals().each_with_index(function(f, i) {
        if (i == 0) {
          return newsrctmp._set(f, target);
        } else {
          param_src = self.$.D._get("src")._get(obj.params()._get(i - 1));
          return newsrctmp._set(f, param_src);
        }
      });
      res = null;
      self.dynamic_bind(function() {
        return res = m.apply(m, [target].concat(params));
      }, new EnsoHash ({ srctemp: newsrctmp }));
      self.$.D._get("src")._set(obj, self.$.D._get("src")._get(clos.body()));
      return res;
    };

    this.eval_EFor = function(obj) {
      var self = this; 
      var env, nenv, res;
      env = new EnsoHash ({ });
      env._set(obj.var(), null);
      nenv = Env.HashEnv.new(env, self.$.D._get("env"));
      res = self.eval(obj.list()).map(function(val) {
        nenv._set(obj.var(), val);
        return self.dynamic_bind(function() {
          return self.eval(obj.body());
        }, new EnsoHash ({ env: nenv }));
      });
      self.$.D._get("src")._set(obj, self.$.D._get("src")._get(obj.body()));
      return res;
    };

    this.eval_InstanceOf = function(obj) {
      var self = this; 
      var a;
      a = self.eval(obj.base());
      return a && Schema.subclass_P(a.schema_class(), obj.class_name());
    };

    this.eval_Eval = function(obj) {
      var self = this; 
      var env1, expr1, res;
      env1 = Env.HashEnv.new();
      obj.envs().map(function(e) {
        return self.eval(e);
      }).each(function(env) {
        if (System.test_type(env, Factory.List)) {
          return env.each(function(v) {
            return env1._set(v.name(), v);
          });
        } else {
          return env.each_pair(function(k, v) {
            return env1._set(k, v);
          });
        }
      });
      expr1 = self.eval(obj.expr());
      res = self.dynamic_bind(function() {
        return self.eval(expr1);
      }, new EnsoHash ({ env: env1 }));
      self.$.D._get("src")._set(obj, self.$.D._get("src")._get(expr1));
      return res;
    };
  });

  Evalexprstencil = {
    EvalExprStencil: EvalExprStencil,

  };
  return Evalexprstencil;
})

File: src/core/diagram/code/evalexprstencil.rb
require 'core/expr/code/impl'
require 'core/expr/code/env'
require 'core/expr/code/env'
require 'core/diagram/code/traceval'
require 'core/schema/code/factory'

module Evalexprstencil

  module EvalExprStencil
    include Traceval::TracevalCommand
#    include Impl::EvalCommand

    def eval_Rule(obj)
      funname = "#{obj.name}__#{obj.type}"
      #create a new function
      forms = [obj.obj]
      obj.formals.each {|f| forms << f.name}
      @D[:env][funname] = Impl::Closure.make_closure(obj.body, forms, @D[:env], self)
      @D[:srctemp][funname] = Impl::Closure.new(obj.body, forms, @D[:env], self)
    end

    def eval_RuleCall(obj)
      target = eval(obj.obj)
      funname = "#{obj.name}__#{target.schema_class.name}"
      m = @D[:env][funname]
      params = obj.params.map{|p|eval(p)}

      clos = @D[:srctemp][funname]
      newsrctmp = @D[:srctemp].clone
      clos.formals.each_with_index do |f,i|
        if (i==0)
          newsrctmp[f] = target
        else
          param_src = @D[:src][obj.params[i-1]]
          newsrctmp[f] = param_src
        end
      end
      res = nil
      dynamic_bind srctemp: newsrctmp do
        res = m.call(target, *params)
      end
      @D[:src][obj] = @D[:src][clos.body]
      res
    end

    def eval_EFor(obj)
      nenv = Env::HashEnv.new({obj.var=>nil}, @D[:env])
      res = eval(obj.list).map do |val|  #returns list of results instead of only last result
        nenv[obj.var] = val
        dynamic_bind env: nenv do
          eval(obj.body)
        end
      end
      @D[:src][obj] = @D[:src][obj.body]
      res
    end
  
    def eval_InstanceOf(obj)
      a = eval(obj.base)
      a && Schema::subclass?(a.schema_class, obj.class_name)
    end

    def eval_Eval(obj)
      env1 = Env::HashEnv.new
      obj.envs.map{|e| eval(e)}.each do |env|
        if env.is_a? Factory::List
          env.each do |v|
            env1[v.name] = v
          end
        else
          env.each_pair do |k,v|
            env1[k] = v
          end
        end
      end
      expr1 = eval(obj.expr)
      res = dynamic_bind env: env1 do
        eval(expr1)
      end
      puts "src for inner=#{@D[:src][expr1]}"
      @D[:src][obj] = @D[:src][expr1]
      res
    end

  end    
end
